lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -0,0 +1,292 @@
+#include "mlir/Support/FileUtilities.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+
+#define DEBUG_TYPE "firrtl-blackbox-reader"
+
+using namespace circt;
+using namespace firrtl;
+
+using hw::OutputFileAttr;
+using sv::VerbatimOp;
+
+/// Append a path to an existing path, replacing it if the other path is
+/// absolute. This mimicks the behaviour of `foo/bar` and `/foo/bar` being used
+/// in a working directory `/home`, resulting in `/home/foo/bar` and `/foo/bar`,
+/// respectively.
+static void appendPossiblyAbsolutePath(SmallVectorImpl<char> &base,
+                                       const Twine &suffix) {
+  if (llvm::sys::path::is_absolute(suffix)) {
+    base.clear();
+    suffix.toVector(base);
+  } else {
+    llvm::sys::path::append(base, suffix);
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct EmittedFile {
+  VerbatimOp op;
+  StringAttr fileName;
+};
+
+struct BlackBoxReaderPass : public BlackBoxReaderBase<BlackBoxReaderPass> {
+  void runOnOperation() override;
+  void runOnAnnotation(Operation *op, DictionaryAttr anno, OpBuilder &builder);
+  void loadFile(Operation *op, StringRef inputPath, OpBuilder &builder);
+
+  using BlackBoxReaderBase::inputPrefix;
+  using BlackBoxReaderBase::resourcePrefix;
+
+private:
+  /// The generated `sv.verbatim` nodes.
+  SmallVector<EmittedFile, 8> emittedFiles;
+
+  /// The target directory to output black boxes into. Can be changed
+  /// through `firrtl.transforms.BlackBoxTargetDirAnno` annotations.
+  SmallString<128> targetDir;
+
+  /// The file list file name (sic) for black boxes. If set, generates a file
+  /// that lists all non-header source files for black boxes. Can be changed
+  /// through `firrtl.transforms.BlackBoxResourceFileNameAnno` annotations.
+  SmallString<128> resourceFileName;
+
+  // Internalized annotation class names for fast comparison.
+  StringAttr targetDirAnnoClass;
+  StringAttr resourceFileNameAnnoClass;
+  StringAttr inlineAnnoClass;
+  StringAttr pathAnnoClass;
+  StringAttr resourceAnnoClass;
+};
+} // end anonymous namespace
+
+/// Emit the annotated source code for black boxes in a circuit.
+void BlackBoxReaderPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  // Internalize some string attributes for easy reference later.
+  auto cx = &getContext();
+  targetDirAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxTargetDirAnno");
+  resourceFileNameAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceFileNameAnno");
+  inlineAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxInlineAnno");
+  pathAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxPathAnno");
+  resourceAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceAnno");
```

Comment:
I hadn't thought to do it this way. Normally, I'll compare the "class" key `StringAttr` _value_ against a `char *`. I assume that by constructing these `StringAttr`'s directly, you can do the comparison by pointer comparison?

Reply:
There is also a benefit of you can then do the `cls == anno.getAs<A>("...")` below which then doesn't require a check before a `getValue` to avoid the null dereference. I'm good with this the way it is. :+1:

Reply:
Yeah exactly, this way it's just a pointer comparison. It's probably a bit esoteric to do this, but it simplifies the check further down as you say.

Reply:
Nice solution. I should steal this pattern for the grand central annotation handling. 😄